docs/rfcs/033-ancestor-deletion.md

```diff
@@ -0,0 +1,237 @@
+
+We could add a layer of indirection that let a child "pretend" that it had no
+ancestor, when in reality it still had the parent. The pageserver API could
+accept deletion of ancestor timelines, and just update child metadata to make
+them look like they have no ancestor.
+
+This would not achieve the desired reduction in storage cost, and may well be more
+complex to maintain than simply implementing the API described in this RFC.
+
+### Avoid copying objects: enable child index to use parent layers directly
+
+We could teach IndexPart to store a TimelineId for each layer, such that a child
+timeline could reference a parent's layers directly, rather than copying them
+into the child's prefix.
+
+This would impose a cost for the normal case of indices that only target the
+timeline's own layers, add complexity, and break the useful simplifying
+invariant that timelines "own" their own path. If child timelines were
+referencing layers from the parent, we would have to ensure that the parent
+never runs GC/compaction again, which would make the API less flexible (the
+proposal in this RFC enables deletion of the parent but doesn't require it.)
+
+## Performance
+
+### Adopting layers
+
+- CopyObject is a relatively cheap operation, but we may need to issue tens of thousands
+  of such requests: this can take up to tens of seconds and will compete for RemoteStorage
+  semaphore units with other activity on the pageserver.
+- If we are running on storage backend that doesn't implement CopyObject, then
+  this part will be much more expensive as we would stream all layer content
+  through the pageserver. This is no different to issuing a lot
+  of reads to a timeline that does not have a warm local cache: it will move
+  a lot of gigabytes, but that shouldn't break anything.
+- Generating truncated layers for delta that straddle the branch point will
+  require streaming read/write of all the layers in question.
+
+### Updating timeline ancestry
+
+The simplest way to update timeline ancestry will probably be to stop and start
+all the Timeline objects: this is preferable to the complexity of making their
+ancestry mutable at runtime.
+
+There will be a corresponding "stutter" in the availability of the timelines,
+of the order 10-100ms, which is the time taken to upload their IndexPart, and
+restart the Timeline.
+
+# Interaction with other features
+
+## Concurrent timeline creation
+
+If new historic timelines are created using the parent as an ancestor while the merge API is running, they will not be re-parented to the child. This doesn't break
+anything, but it leave the parent in a state where it might not be possible
+to delete it.
+
+Since timeline creations are an explicit user action, this is not something we need to
+worry about as the storage layer: a user who wants to delete their parent timeline will not create
+new children, and if they do, they can choose to delete those children to
+enable deleting the parent.
+
```

Comment:
For consumption metrics, if we'd just copy the logical size cache while when unlinking the "old main" and "new main", we would get disruptionless synthetic size.

The synthetic size will then increase a lot temporarily because of multiple root timelines. I think that is fine. Or has there been any discussion that it shouldn't increase? If there should not be any billed cost we could have a "weak_ancestor" on the "new main" pointing to "old main", and reducing the cost of the duplication operation.

Reply:
Interesting to think about.  They will indeed have a moment when the synthetic size ~doubles, before some higher layer deletes the old main timeline.

My bet is that since we charge per `size/time`, so a short spike probably won't show up in a way that upsets anyone, and we're kind of accurately passing along our costs to store both timelines for a moment.  Still, it would be kind of unfair if the charge appears for some users and not for others based on whether a consumption metrics poll hits the system at the right moment.

I think our starting point can be to do no special handling, and if we wanted to later change that we could add a flag to timelines to disable counting them for consumption metrics, and add an option to the merge API that would automatically set the "free" flag on the parent.  Something to pick up when integrating this API with cloud control plane.